docs/root/configuration/other_features/internal_listener.rst

```diff
@@ -0,0 +1,100 @@
+.. _config_internal_listener:
+
+Internal Listener
+=================
+
+How it works
+------------
+
+This extension contains 2 major components to add a listener with
+an :ref:`envoy internal address <envoy_v3_api_msg_config.core.v3.EnvoyInternalAddress>`
+and to create a client connection to that :ref:`listener <envoy_v3_api_msg_config.listener.v3.Listener>`
+
+**envoy.bootstrap.internal_listener_registry** This bootstrap extension is required to support looking up the target listener via an
+:ref:`envoy internal address <envoy_v3_api_msg_config.core.v3.EnvoyInternalAddress>` on each worker threads.
+
+**network.connection.client.envoy_internal** It is a client connection factory. The factory is implicitly instantiated by the dispatcher to establish a client connection to an
+internal listener address.
+
+Example config
+--------------
+Below is a smallest static config that redirect tcp_proxy on port 19000 to the tcp_proxy binding to the internal address.
+
+.. code-block:: yaml
```

Comment:
Best to switch to verifiable config fragments, @phlax what is the current best practice for this?

Reply:
Please address this one and then API LGTM>

Reply:
Can I start put the yaml in config/ directory and link the yaml like https://github.com/envoyproxy/envoy/blob/main/docs/root/intro/arch_overview/http/upgrades.rst#tunneling-tcp-over-http ?

And add the verifiable config as follow up?

Reply:
i think best just add the config initially, but ill leave it to you to decide

to do so do

```console
git grep literalinclude docs/
```
you should find a few existing examples

Reply:
done

Reply:
I was almost there. The yaml is magically tested. However, I am seeing a failure related to a runtime guard. It can be reproduced at local. 
This complains even if I explicitly enable it here https://github.com/lambdai/envoy-dai/blob/intlisbootstrapext/docs/root/configuration/other_features/_include/internal-listener-config.yaml#L62
```
[2022-03-10 22:15:15.856][12][critical][assert] [source/server/listener_manager_impl.cc:339] assert failure: !config.address().has_envoy_internal_address(). Details: listener singleton_internal_encap has envoy internal address server_listener_name: "singleton_internal_encap"
. This runtime feature is disabled.
```

Reply:
@phlax So the internal-listener-config.yaml can be loaded by buildout envoy but not in config_test

Is there anything I can specialize for internal-listener-config.yaml ?

Reply:
not sure - usually the only time it doesnt work with config_test is if there are some resources - ie certs/wasm/etc - that need to be loaded - and then its a matter of ensuring that it points to available resources.

im not so familiar with how `config_test` itself works, altho my feeling is it should work similarly enough to the built binary that configs should be compatible without modification

that it doesnt suggests to me that its not the same binary being used, altho i confess i also dont know too much about runtime guarding either

Reply:
im wondering if its possible that config_test cant handle `bootstrap_extension` ?

i can see one of the example yaml files in the docs has a `layered_runtime` but i dont see any with `bootstrap_extension` so im wondering if that is the culprit

Reply:
I don't think there is explicit support for bootstrap extension, but there is a fair bit of mocking going on there, so it's possible some dependency needs adding or tuning.

Reply:
@htuch @phlax 
Can I revert this verified_config and address that later? IMO the efforts is far off the topic of this PR

I can explore in another dedicated one and it would be easier to review

Reply:
I can help with a verified fragment later. We probably need an end-to-end sample as well. This PR is blocking implementation for #19435.

Reply:
Updating...

Reply:
create https://github.com/envoyproxy/envoy/issues/20335 to track the next step